content/post/2019-10-07-Image-Creation-Process.md

```diff
@@ -0,0 +1,53 @@
+
+## Stage 1: Building a Base, Ex Nihilo
+
+Image creation starts with building a base.  This is done by downloading an install ISO for a linux distribution, or a prepared image such as a virtual network appliance.  
+
+We boot this image to add the “antidote” user and add it to the appropriate user groups.  In most cases we add this user to the “sudo” group, or otherwise give this user equivalent administrative privileges.  For instance, in the Cumulus VX image, we add the user to the “netshow” and “netedit” groups so it can run Cumulus’s “net” tool.  Next, we ensure that the primary interface of the image is configured for DHCP.  
+
+In the case of applications like Asterisk PBX, Free-Range Routing, or Netbox images, we install or compile the relevant packages and dependencies.  On all images, we ensure that any APIs (including REST or Netconf) are enabled.
+
+Once we are done with the ISO image itself, we need to create a dockerfile for the image that launches it with QEMU.  The launch script for this container should look for and create the right network interfaces with QEMU, and stitch them to the underlying interfaces in the container.  It should also pass a mount-point to the virtual-machine so it can mount the lesson and configuration files.  Finally, the launch script should ensure the right ports are being forwarded from the underlying primary interface of the container (“eth0”) to the primary interface of the virtual-machine.
+
+The completed ISO files for QEMU are stored in a Google Cloud Platform bucket.  The docker images are stored in the “antidotelabs” docker hub repository.  Our base is done!
+
+## Stage 2:  The Lesson Metafile
+
+Now that we have a base image, we need to start configuring it for a particular lesson at the time the lesson is launched.  This configuration starts with Syringe, the core component of the Antidote platform, reading the lesson metafile at the time a lesson is launched.
+
+This Metafile describes which images to use for the lesson, and how the user will interact with those images.  For instance, the user might use SSH, NetConf, or a REST API to interact with the image.  
+
+Additionally, the metafile describes how the various virtual-machines in a lesson are interconnected, and Syringe will ensure that the right interfaces are created and connected with each other when the virtual-machines are launched.
+
+So far so good.  The base images are launched, we can access them, and they are wired together!
+
+## Stage 3:  Virtual-Machine Configuration
+
+The final step in preparing the image for the user is to configure the applications in it.  This can be done a variety of ways, including via NetConf, Python, or Ansible.  First, though, this is the stage where the hostname and interface IP addresses of the virtual-machine are configured.  
```

Comment:
```suggestion
The final step in preparing the image for the user is to configure the applications in it. [This can be done a variety of ways](https://antidoteproject.readthedocs.io/en/latest/platform/curricula/lessons/configuration.html), including via NetConf, Python, or Ansible.  First, though, this is the stage where the hostname and interface IP addresses of the virtual-machine are configured.  
```